geoSegment.js

```diff
@@ -0,0 +1,376 @@
+const SCHEDULE_TYPE = 'Schedule';
+const MINIMAL_EVENT_SCHEDULE_SCHEMA = Joi.object().keys({
+  // If this array has length 0, an infinite loop will occur in the date
+  // generation algorithm below.
+  byDay: Joi.array().items(Joi.string()).min(1).required(),
+  startTime: Joi.string().required(),
+  endTime: Joi.string().required(),
+  idTemplate: Joi.string().required(),
+});
+
+/**
+ * @typedef {import('axios').AxiosError} AxiosError
+ * @typedef {import('axios').AxiosRequestConfig} AxiosRequestConfig
+ * @typedef {import('axios').AxiosResponse} AxiosResponse
+ * @typedef {{
+ * location?: {
+ *  geo?: {
+ *   latitude: number,
+ *   longitude: number,
+ *   }
+ *  },
+ * [k:string]: unknown
+ * }} SessionSeriesData
+ * @typedef {{
+ *  id: string,
+ *  state: 'updated' | 'deleted',
+ *  kind: 'SessionSeries',
+ *  data: SessionSeriesData
+ * }} SessionSeriesItem
+ * @typedef {{
+ *  startDate?: string,
+ *  superEvent?: string,
+ *  [k:string]: unknown
+ * }} ScheduledSessionData
+ * @typedef {{
+ *  id: string,
+ *  state: 'updated' | 'deleted',
+ *  kind: 'ScheduledSession',
+ *  data: ScheduledSessionData
+ * }} ScheduledSessionItem
+ * @typedef {{
+ *  latitude: number,
+ *  longitude: number,
+ *  radius: number,
+ *  identifier: string
+ * }} Segment
+ */
+
+/**
+ * @param {string} segmentIdentifier
+ */
+function getSegmentIndexFilePath(segmentIdentifier) {
+  return `${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/index.txt`;
+}
+
+/**
+ * @param {string} sessionSeriesId
+ */
+function getSessionSeriesFilePath(sessionSeriesIdHash) {
+  return `${SESSION_SERIES_DIRECTORY_PATH}/${sessionSeriesIdHash}.json`;
+}
+
+/**
+ * @param {string} segmentIdentifier
+ * @param {string} scheduledSessionIdHash
+ */
+function getScheduledSessionFilePath(segmentIdentifier, scheduledSessionIdHash) {
+  return `${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/${scheduledSessionIdHash}.json`;
+}
+
+/**
+ * @param {Segment[]} segments
+ */
+async function createSegmentDirectories(segments) {
+  // Create segment directories;
+  segments.map((segment) => segment.identifier)
+    .map(async (segmentIdentifier) => {
+      const segmentDirectoryPath = `${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}`;
+      await emptyOrMakeDirectory(segmentDirectoryPath);
+
+      // Create index file
+      await fsPromises.writeFile(getSegmentIndexFilePath(segmentIdentifier), '');
+    });
+}
+
+/**
+ * @param {ScheduledSessionData} scheduledSessionData
+ * @param {SessionSeriesData} sessionSeries
+ */
+async function linkScheduledSessionAndSessionSeriesAndWrite(scheduledSessionData, sessionSeries) {
+  // Link ScheduledSession with SessionSeries
+  const linkedScheduledSessionData = {
+    ...scheduledSessionData,
+    superEvent: sessionSeries,
+  };
+
+  // Write ScheduledSession into each output segment directory
+  for (const segmentIdentifier of sessionSeries['imin:segment']) {
+    const scheduledSessionIdHash = hashString(linkedScheduledSessionData.id);
```

Comment:
Also can I just check - is every ScheduledSession in our platform 100% guaranteed to have an `id`, or do we need to fall back to the RPDE id if there's no id available?

Reply:
@nickevansuk that is how Firehose works currently. All items will have an ID. It looks like Civ's put a check in `processScheduledSessionItems` which means that ScheduledSessions without IDs will be filtered out (just in case Firehose's id contract changes in the future)

Reply:
Ok great 👍